operator/beiboot/handler/beiboots.py

```diff
@@ -1,266 +1,104 @@
-        logger.info(f"Deleted namespace for beiboot: {namespace}")
-    except k8s.client.exceptions.ApiException:
-        pass
+from beiboot.clusterstate import BeibootCluster
 
 
+@kopf.on.resume("beiboot")
 @kopf.on.create("beiboot")
 async def beiboot_created(body, logger, **kwargs):
-    provider = body.get("provider")  # k3s
-    name = body["metadata"]["name"]
-    ports = body.get("ports")
-
-    cluster_config = configuration.refresh_k8s_config()
-    #
-    # Create the target namespace
-    #
-    try:
-        namespace = handle_create_namespace(logger, name, cluster_config)
-
-        if provider == "k3s":
-            node_token = generate_token()
-            cgroup = "".join(e for e in name if e.isalnum())
-            server_workloads = [
-                create_k3s_server_workload(
-                    namespace, node_token, cgroup, cluster_config
-                )
-            ]
-            node_workloads = [
-                create_k3s_agent_workload(
-                    namespace, node_token, cgroup, cluster_config, node
-                )
-                for node in range(
-                    1, cluster_config.nodes
-                )  # (no +1 ) since the server deployment already runs one node
-            ]
-            services = [create_k3s_kubeapi_service(namespace, cluster_config)]
-        else:
-            raise kopf.PermanentError(
-                f"Cannot create Beiboot wit provider {provider}: not supported."
+    """
+    > If the cluster is not running, try to create it. If it is running, reconcile it
+
+    :param body: The body of the Kubernetes resource that triggered the event
+    :param logger: the logger object
+    """
+    parameters = configuration.refresh_k8s_config()
+    cluster = BeibootCluster(configuration, parameters, model=body, logger=logger)
+
+    if cluster.is_requested or cluster.is_creating:
+        # this is the initial process for a Beiboot
+        try:
+            if cluster.is_requested:
+                await cluster.create()
+            if cluster.is_creating:
+                await cluster.boot()
+        except kopf.PermanentError as e:
+            await cluster.impair(str(e))
+            raise e from None
+        except Exception as e:  # noqa
+            logger.error(traceback.format_exc())
+            logger.error(
+                "Could not create cluster due to the following error: " + str(e)
             )
-
-        # check if beiboot exists, handle that case
-
-        #
-        # ports to service
-        #
-        additional_services = ports_to_services(ports, namespace, cluster_config)
-        if additional_services:
-            services.extend(additional_services)
-        #
-        # Gefyra Nodeport
-        #
-        if hasattr(cluster_config, "gefyra") and cluster_config.gefyra.get("enabled"):
+            await cluster.impair(str(e))
+            raise kopf.PermanentError(str(e))
+
+    if cluster.is_pending:
```

Comment:
Why not `elif`?

Reply:
if the cluster was first in requested or creating state I want it to fall in this branch afterwards.